data_model/primitives/src/conststr.rs

```diff
@@ -0,0 +1,670 @@
+    ops::Deref,
+    slice::from_raw_parts,
+};
+#[cfg(feature = "std")]
+use std::str::from_utf8_unchecked;
+
+use derive_more::{AsRef, DebugCustom, Display, From};
+use iroha_schema::{IntoSchema, MetaMap};
+use parity_scale_codec::WrapperTypeEncode;
+use serde::{Serialize, Serializer};
+
+const MAX_INLINED_STRING_LEN: usize = 2 * size_of::<usize>() - 1;
+
+/// Immutable inlinable string.
+/// Strings shorter than 15/7/3 bytes (in 64/32/16-bit architecture) are inlined.
+#[derive(Clone, Display, DebugCustom, From, AsRef)]
+#[from(forward)]
+#[as_ref(forward)]
+#[display(fmt = "{}", "&**self")]
+#[debug(fmt = "{:?}", "&**self")]
+pub struct ConstString(ConstStringData);
+
+impl ConstString {
+    /// Returns the length of this [`ConstString`], in bytes.
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    /// Returns `true` if [`ConstString`] is empty.
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.0.len() == 0
+    }
+
+    /// Returns `true` if [`ConstString`] data is inlined.
+    #[inline]
+    pub const fn is_inlined(&self) -> bool {
+        self.0.is_inlined()
+    }
+
+    /// Construct empty [`ConstString`].
+    #[inline]
+    pub const fn new() -> Self {
+        Self(ConstStringData::new())
+    }
+}
+
+impl Deref for ConstString {
+    type Target = str;
+
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        self.as_ref()
+    }
+}
+
+impl Hash for ConstString {
+    #[inline]
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        (**self).hash(state)
+    }
+}
+
+impl Ord for ConstString {
+    #[inline]
+    fn cmp(&self, other: &Self) -> Ordering {
+        Ord::cmp(&**self, &**other)
+    }
+}
+
+/// Can't be derived.
+impl PartialOrd for ConstString {
+    #[inline]
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl PartialEq for ConstString {
+    #[inline]
+    fn eq(&self, other: &Self) -> bool {
+        PartialEq::eq(&**self, &**other)
+    }
+}
+
+macro_rules! impl_eq {
+    ($($ty:ty,)*) => {
+        impl_eq!($($ty),*);
+    };
+    ($($ty:ty),*) => {$(
+        impl PartialEq<$ty> for ConstString {
+            // Not possible to write macro uniformly for different types otherwise.
+            #[allow(clippy::string_slice, clippy::deref_by_slicing)]
+            #[inline]
+            fn eq(&self, other: &$ty) -> bool {
+                PartialEq::eq(&self[..], &other[..])
+            }
+        }
+
+        impl PartialEq<ConstString> for $ty {
+            // Not possible to write macro uniformly for different types otherwise.
+            #[allow(clippy::string_slice, clippy::deref_by_slicing)]
+            #[inline]
+            fn eq(&self, other: &ConstString) -> bool {
+                PartialEq::eq(&self[..], &other[..])
+            }
+        }
+    )*};
+}
+
+impl_eq!(String, str, &str);
+
+/// Can't be derived.
+impl Eq for ConstString {}
+
+impl Serialize for ConstString {
```

Comment:
how come there is no deserialze implementation?

Reply:
`Name` struct doesn't have `Deserialize` and `Decode`, so i decided not to implement them for now.
I can implement then if needed.

Reply:
It has them implemented manually, no?

Reply:
oh, i missed them...
I will add `Decode` and `Deserialize` for `ConstString`!